### PR TITLE
[Inductor] Guard loop reindexing against coalescing loss

### LINKED ISSUE #189488: [Inductor][Convolution] #183521 regresses conv models without `TORCHINDUCTOR_LOOP_REINDEXING_AFTER_FUSION`
### 🐛 Describe the bug

#183521 appears to have introduced a regression in models such as Stable Diffusion due to changes in conv fusions. Minimal repro attached

bad
```
latency_ms=mean=122.874 p50=122.677 p90=123.610 min=122.045
```
good
```
latency_ms=mean=20.421 p50=20.430 p90=20.464 min=20.313
```

repro commands:

```
  TORCHINDUCTOR_CACHE_DIR=/tmp/ti_repro_on \
  python minimal_inductor_repro.py \
    --variant residual_upsample_conv_skipconv \
    --channels 256 \
    --input-size 128 \
    --iters 6 \
    --warmup 2
```

```
  TORCHINDUCTOR_CACHE_DIR=/tmp/ti_repro_off \
  TORCHINDUCTOR_LOOP_REINDEXING_AFTER_FUSION=0 \
  python minimal_inductor_repro.py \
    --variant residual_upsample_conv_skipconv \
    --channels 256 \
    --input-size 128 \
    --iters 6 \
    --warmup 2
```

```    
#!/usr/bin/env python3
import argparse
import os
import statistics
import time

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config, metrics


class UpsampleConvGroupNormSilu(nn.Module):
    def __init__(self, channels: int, groups: int) -> None:
        super().__init__()
        self.conv = nn.Conv2d(channels, channels, kernel_size=3, padding=1)
        self.norm = nn.GroupNorm(groups, channels, eps=1e-6)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = F.interpolate(x, scale_factor=2.0, mode="nearest")
        x = self.conv(x)
        x = self.norm(x)
        return F.silu(x)


class UpsampleConvGroupNormSiluResidual(nn.Module):
    def __init__(self, channels: int, groups: int) -> None:
        super().__init__()
        self.conv = nn.Conv2d(channels, channels, kernel_size=3, padding=1)
        self.norm = nn.GroupNorm(groups, channels, eps=1e-6)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = F.interpolate(x, scale_factor=2.0, mode="nearest")
        x = self.conv(x)
        y = F.silu(self.norm(x))
        return y + (x * 0.01)


class UpsampleConvGroupNormSiluSkipConv(nn.Module):
    def __init__(self, channels: int, groups: int) -> None:
        super().__init__()
        self.conv = nn.Conv2d(channels, channels, kernel_size=3, padding=1)
        self.norm = nn.GroupNorm(groups, channels, eps=1e-6)
        self.skip = nn.Conv2d(channels, channels, kernel_size=1)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = F.interpolate(x, scale_factor=2.0, mode="nearest")
        x = self.conv(x)
        y = F.silu(self.norm(x))
        skip = self.skip(x)
        return y + (skip * 0.01)


class ResidualUpsampleConvGroupNormSiluSkipConv(nn.Module):
    def __init__(self, channels: int, groups: int) -> None:
        super().__init__()
        self.conv = nn.Conv2d(channels, channels, kernel_size=3, padding=1)
        self.norm = nn.GroupNorm(groups, channels, eps=1e-6)
        self.skip = nn.Conv2d(channels, channels, kernel_size=1)

    def forward(self, x: torch.Tensor, residual: torch.Tensor) -> torch.Tensor:
        x = (x + residual) / 1.41421356237
        x = F.interpolate(x, scale_factor=2.0, mode="nearest")
        x = self.conv(x)
        y = F.silu(self.norm(x))
        skip = self.skip(x)
        return y + (skip * 0.01)


class UpsampleGroupNormSilu(nn.Module):
    def __init__(self, channels: int, groups: int) -> None:
        super().__init__()
        self.norm = nn.GroupNorm(groups, channels, eps=1e-6)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = F.interpolate(x, scale_factor=2.0, mode="nearest")
        x = self.norm(x)
        return F.silu(x)


class GroupNormSilu(nn.Module):
    def __init__(self, channels: int, groups: int) -> None:
        super().__init__()
        self.norm = nn.GroupNorm(groups, channels, eps=1e-6)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.norm(x)
        return F.silu(x)


def percentile(values: list[float], pct: float) -> float:
    if not values:
        return float("nan")
    ordered = sorted(values)
    idx = round((len(ordered) - 1) * pct / 100.0)
    return ordered[idx]


def synchronize(device: torch.device) -> None:
    if device.type == "cuda":
        torch.cuda.synchronize(device)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--batch", type=int, default=64)
    parser.add_argument("--channels", type=int, default=256)
    parser.add_argument("--groups", type=int, default=32)
    parser.add_argument("--input-size", type=int, default=8)
    parser.add_argument(
        "--variant",
        choices=(
            "upsample_conv",
            "upsample_conv_residual",
            "upsample_conv_skipconv",
            "residual_upsample_conv_skipconv",
            "upsample_norm",
            "norm",
        ),
        default="upsample_norm",
    )
    parser.add_argument("--channels-last", action="store_true")
    parser.add_argument("--warmup", type=int, default=5)
    parser.add_argument("--iters", type=int, default=30)
    parser.add_argument("--eager", action="store_true")
    args = parser.parse_args()

    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required for this reproducer")

    torch.manual_seed(0)
    torch.backends.cuda.matmul.allow_tf32 = True
    torch.backends.cudnn.allow_tf32 = True

    device = torch.device("cuda")
    if args.variant == "upsample_conv":
        model = UpsampleConvGroupNormSilu(args.channels, args.groups)
    elif args.variant == "upsample_conv_residual":
        model = UpsampleConvGroupNormSiluResidual(args.channels, args.groups)
    elif args.variant == "upsample_conv_skipconv":
        model = UpsampleConvGroupNormSiluSkipConv(args.channels, args.groups)
    elif args.variant == "residual_upsample_conv_skipconv":
        model = ResidualUpsampleConvGroupNormSiluSkipConv(args.channels, args.groups)
    elif args.variant == "upsample_norm":
        model = UpsampleGroupNormSilu(args.channels, args.groups)
    else:
        model = GroupNormSilu(args.channels, args.groups)

    model = model.eval().to(device)
    if args.channels_last:
        model = model.to(memory_format=torch.channels_last)

    x = torch.randn(
        args.batch,
        args.channels,
        args.input_size,
        args.input_size,
        device=device,
    )
    if args.channels_last:
        x = x.to(memory_format=torch.channels_last)
    residual = torch.randn_like(x) if args.variant.startswith("residual_") else None

    compiled = model if args.eager else torch.compile(model, mode="reduce-overhead")

    metrics.reset()
    with torch.no_grad():
        for _ in range(args.warmup):
            y = compiled(x, residual) if residual is not None else compiled(x)
        synchronize(device)
        generated_kernel_count = metrics.generated_kernel_count

        times_ms = []
        for _ in range(args.iters):
            start = time.perf_counter()
            y = compiled(x, residual) if residual is not None else compiled(x)
            synchronize(device)
            times_ms.append((time.perf_counter() - start) * 1000.0)

    # Keep the output live without adding work to the timed loop.
    checksum = float(y.mean().detach().cpu())

    print(f"torch={torch.__version__}")
    print(f"device={torch.cuda.get_device_name(device)}")
    print(
        "loop_reindexing_after_fusion="
        f"{os.environ.get('TORCHINDUCTOR_LOOP_REINDEXING_AFTER_FUSION', '<default>')}"
    )
    print(
        "inductor_config="
        f"loop_reindexing_after_fusion={config.loop_reindexing_after_fusion} "
        f"loop_ordering_after_fusion={config.loop_ordering_after_fusion}"
    )
    print(
        "shape="
        f"input=({args.batch}, {args.channels}, {args.input_size}, {args.input_size}) "
        f"output=({args.batch}, {args.channels}, "
        f"{args.input_size * (1 if args.variant == 'norm' else 2)}, "
        f"{args.input_size * (1 if args.variant == 'norm' else 2)}) "
        f"groups={args.groups} variant={args.variant} "
        f"channels_last={args.channels_last}"
    )
    print(f"generated_kernel_count={generated_kernel_count}")
    print(
        "latency_ms="
        f"mean={statistics.mean(times_ms):.3f} "
        f"p50={statistics.median(times_ms):.3f} "
        f"p90={percentile(times_ms, 90):.3f} "
        f"min={min(times_ms):.3f}"
    )
    print(f"checksum={checksum:.6f}")


if __name__ == "__main__":
    main()
```

### Versions

builds containing #183521

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by guan404ming**
## Related Issue

Fixes #189488

## Why

- Reindexing rewrites a pointwise's loops to match a reduction's split
- It only checks deps match, never what coalescing it costs
- Upsample into GroupNorm lost coalescing, running 6x slower

## How

- Measure uncoalesced memory the reindex adds, roll back when it exceeds the fusion's savings
- Skip the guard when the coalesce analysis is unavailable on either side, since `None` is not a score of 0
- Log rollbacks to `loop_ordering` so the decision is visible in a `TORCH_LOGS` dump

On the threshold: `score_fusion_memory()` is 0 when deps match only after normalization, so the guard is zero-tolerance in that case and proportional elsewhere. Both are load-bearing. Replacing it with the normalized-matched shared bytes lets #189488 regress (6 kernels back to 5), and hardcoding zero tolerance breaks `test_reshape_reindexing_transposed_input`.

## Before / after

Issue repro on RTX 5090, `--variant residual_upsample_conv_skipconv --channels 256 --input-size 128`:

| | latency | kernels |
|---|---|---|
| main | 592.4ms | 4 |
| main + `TORCHINDUCTOR_LOOP_REINDEXING_AFTER_FUSION=0` | 102.5ms | 6 |
| this PR | 102.4ms | 6 |

Output checksum is identical in all three.

Compile time on the regression test's model, best of 3 with a fresh cache: 0.93s on main, 1.19s here. The delta covers both the coalesce analysis and the different fusion outcome.

## Test

```
python test/inductor/test_loop_ordering.py
python test/inductor/test_perf.py
```

101 passed on an RTX 5090.

This PR was authored with the assistance of Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by guan404ming** [comment]
Kept the threshold: both suggested alternatives regress. Added None handling and logging.

**Comment by ezyang** [comment]
@claude fix seems pretty ugly ngl

**Comment by guan404ming** [comment]
Renamed to `coalescing_budget` and made the zero-budget case a documented intent rather than a side effect of `score_fusion_memory`'s exact intersection. Also measured dropping the before term: `uncoalesced_before` is nonzero in `test_keep_fake_dep` (256), so it stays. 
 
 Both test files green on RTX 5090. The mutate-measure-rollback structure itself would need the `_FusedNodeView` rework, which I think it would be better to leave to a follow-up. Hopefully cleaner now and please feel free to let me know where could be better, thanks!

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -7542,6 +7554,22 @@ def _try_reindex_pointwise_for_reduction(
             rollback_snapshot.restore()
             return False
 
+        # Budget is 0 by design when deps only match after normalize():
+        # that weaker win must not cost any coalescing.
+        uncoalesced_after = _uncoalesced_memory_score(snodes)
+        if uncoalesced_before is not None and uncoalesced_after is not None:
+            uncoalesced_added = uncoalesced_after - uncoalesced_before
+            coalescing_budget = self.score_fusion_memory(node1, node2)
```
Comment (eellison) [inline torch/_inductor/scheduler.py:7562]: Can we compare the sum of coalesced & uncoalesced memory for node1, node2, and the potential fused node1-node2? We can weight the uncoalesced memory by a large penalty.  

#189488 was about 6x slower despite the static fused-view score looking like roughly 8 GiB coalesced / 4 GiB uncoalesced versus roughly 20 GiB coalesced / 0 uncoalesced for the individual nodes

You can use the fused view with (`analyze_memory_coalescing_for_nodes([node1, node2])`)

**Review by eellison — COMMENTED** [review]
good start, but, i think we can improve it a bit

File: torch/_inductor/scheduler.py (hunk 1 above)
Comment (guan404ming) [reply torch/_inductor/scheduler.py:7562]: Thanks for the suggestion. Tried it on the repro (main 592ms/4k, PR 102ms/6k); kept the guard. Findings:

1. `analyze_memory_coalescing_for_nodes([node1, node2])` returns empty analysis when either is a `FusedSchedulerNode`: `extract_normalized_read_writes` skips non-`SchedulerNode` members of the view. Flattening to leaves works around it.
2. The fused view reports zero `uncoalesced_addrs`: after its normalization the harmful upsample read becomes a top-level term, so it lands in `coalesced_by_var` under a secondary var. Both bad fusions pass, 529ms.
3. Weighting only the best var rejects the first fusion by a 0.002% margin and still accepts the second: 5 kernels, 529ms.
4. Accept side is the binding constraint: `test_reshape_reindexing_transposed_input` adds 262144 uncoalesced against 266240 shared and must fuse; any P>1 penalty formula rejects it.
5. So the workable comparison degenerates to `uncoalesced_added <= shared_bytes`, which is what the guard already does. On (2), `coalesced_by_var` entries are per-var conditional potentials, so summing them across vars overstates the fused kernel; the silent empty result in (1) still looks like an upstream footgun.

Comment (eellison) [reply torch/_inductor/scheduler.py:7562]: > analyze_memory_coalescing_for_nodes([node1, node2]) returns empty analysis when either is a FusedSchedulerNode: extract_normalized_read_writes skips non-SchedulerNode members of the view. Flattening to leaves works around it.

this is a limitation today, but not a fundamental limitation. do you agree we should be comparing the total coalesced + uncoalesced memory of the individual nodes compared to the fused node? can we do that ?

Comment (guan404ming) [reply torch/_inductor/scheduler.py:7562]: Yes, agreed, and yes we can do that once the fused-view analysis is fixed (flattening + the normalization reclassifying the harmful reads). 

One data point from this repro: the penalty can't be large, since test_reshape_reindexing_transposed_input must still fuse (+262144 uncoalesced vs 266240 shared), which pins P <= ~2 and makes the comparison reduce to the current guard. 

So I think to keep this guard for now and do the full individual-vs-fused comparison as a follow-up on top of the fused-view fix would be a great option. Feel free to let me know if there is any great idea, thanks!

Comment (eellison) [reply torch/_inductor/scheduler.py:7562]: I am going to try to fix in https://github.com/pytorch/pytorch/pull/191349

The scheduler / loop manipulation can be a bit tricky and not necessarily the best for new contributor, sorry !

Comment (guan404ming) [reply torch/_inductor/scheduler.py:7562]: Sure, feel free to take over this. Happy to learn more from your PR, thanks!
